{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d0b504d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml as yml\n",
    "import pandas as pd\n",
    "import math\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c45a7117-8e0a-4dbb-95b3-dbb64257087f",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"config/config.yaml\", \"r\") as f:\n",
    "    config = yml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5656b765-8314-4316-b806-4a24a2105bb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_meta = pd.read_csv(config[\"dna_paired_samples_tsv\"], sep=\"\\t\", header=None, comment=\"#\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c159fbf2-94ae-41b2-a7f1-5676a9b02232",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_meta = pd.read_csv(config[\"rna_paired_samples_tsv\"], sep=\"\\t\", header=None, comment=\"#\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4234a93-de1d-4ef5-99d0-6e1f50eab985",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a list of metrics files that Dragen generates as part of the Tamor pipeline for somatic dna and rna analysis\n",
    "\n",
    "# DNA germline, includes some files generated for the normal during paired tumor-normal analysis. Excludes some duplicate\n",
    "# file generated by same (e.g. .dna.somatic.wgs_contig_mean_cov_normal.csv is same as .dna.germline.wgs_overall_mean_cov.csv).\n",
    "dna_germline_csv_metric_file_suffices = [\n",
    "                                     \".dna.germline.ploidy_estimation_metrics.csv\",\n",
    "                                     \".dna.germline.roh_metrics.csv\",\n",
    "                                     \".dna.germline.time_metrics.csv\", \n",
    "                                     \".dna.germline.vc_hethom_ratio_metrics.csv\", \n",
    "                                     \".dna.germline.vc_metrics.csv\", \n",
    "                                     \".dna.germline.wgs_coverage_metrics.csv\", \n",
    "                                     \".dna.germline.wgs_hist.csv\",\n",
    "                                     \".dna.germline.wgs_overall_mean_cov.csv\",\n",
    "                                     \".dna.germline.sv_metrics.csv\",\n",
    "                                     \".dna.germline.trimmer_metrics.csv\",\n",
    "#                                     \".dna.germline.fastqc_metrics.csv\",\n",
    "                                     \".dna.germline.cnv_metrics.csv\",\n",
    "                                     \".dna.germline.mapping_metrics.csv\",\n",
    "                                     \".dna.somatic.hla_metrics.csv\"]\n",
    "\n",
    "# These germline-related files will have a different prefix since they are generated during the somatic analysis.\n",
    "dna_germline_normal_csv_metric_file_suffices = [\n",
    "                                     \".dna.somatic.tmb_overall_mean_cov_normal.csv\",\n",
    "                                     \".dna.somatic.tmb_coverage_metrics_normal.csv\"]\n",
    "\n",
    "# Single out some metrics files for special processing because they have very large dimensions, but only some values need to be retained, \n",
    "# e.g. per-contig metrics that include all the decoys, alt contigs, etc. but we will just report the reference autosomal, sex and mito contigs\n",
    "dna_germline_contig_csv_metric_file_suffices = [\".dna.germline.wgs_contig_mean_cov.csv\"]\n",
    "\n",
    "#dna_germline_contig_normal_csv_metric_file_suffices = [\".dna.somatic.tmb_contig_mean_cov_normal.csv\"]\n",
    "                                            \n",
    "# Note that we are not capturing histogram bin metrics files as these cause major column bloat, e.g. \".dna.germline.fragment_length_hist.csv\"\n",
    "dna_somatic_csv_metric_file_suffices = [\".dna.somatic.cnv_metrics.csv\", \n",
    "                                        \".dna.somatic.ploidy_estimation_metrics.csv\",\n",
    "                                        \".dna.somatic.sv_metrics.csv\",\n",
    "                                        \".dna.somatic.time_metrics.csv\",\n",
    "                                        \".dna.somatic.vc_metrics.csv\",\n",
    "                                        \".dna.somatic.wgs_overall_mean_cov_tumor.csv\",\n",
    "                                        \".dna.somatic.wgs_coverage_metrics_tumor.csv\",\n",
    "                                        \".dna.somatic.tmb_overall_mean_cov_tumor.csv\",\n",
    "                                        \".dna.somatic.tmb_coverage_metrics_tumor.csv\",\n",
    "                                        \".dna.somatic.hla_metrics.tumor.csv\",\n",
    "                                        \".dna.somatic.tmb.metrics.csv\",\n",
    "                                        \".dna.somatic.mapping_metrics.csv\", \n",
    "#                                         \".dna.somatic.fastqc_metrics.csv\",\n",
    "                                        \".dna.somatic.trimmer_metrics.csv\"]\n",
    "\n",
    "dna_somatic_contig_csv_metric_file_suffices = [\".dna.somatic.wgs_contig_mean_cov_tumor.csv\"]\n",
    "#                                               \".dna.somatic.tmb_contig_mean_cov_tumor.csv\"]\n",
    "\n",
    "# The following files require special procesing as they have their own format: \n",
    "# .dna.germline.insert-stats.tab, \n",
    "# .dna.somatic.insert-stats.tab, \n",
    "# .dna.somatic.hrdscore.csv,\n",
    "# .dna.somatic.microsat_output.json, \n",
    "# and .dna.somatic.cnv.vcf.gz (for ModelSource=DEGENERATE_DIPLOID, etc.)\n",
    "\n",
    "rna_csv_metric_file_suffices = [\".rna.fusion_metrics.csv\",\n",
    "#                                \".rna.fastqc_metrics.csv\",\n",
    "                                \".rna.mapping_metrics.csv\",\n",
    "                                \".rna.wgs_overall_mean_cov.csv\",\n",
    "                                \".rna.wgs_coverage_metrics.csv\",\n",
    "                                \".rna.wgs_coverage_metrics.csv\",\n",
    "                                \".rna.trimmer_metrics.csv\",\n",
    "                                \".rna.time_metrics.csv\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75ad7b8f-1642-4c71-8b15-d6b4cae15bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "metrics = []\n",
    "hhmmss_regex = re.compile(r'(\\d\\d):(\\d\\d):(\\d\\d)')\n",
    "canonical_chr_regex = re.compile(r'^chr(\\d+|X|Y|M|EBV)$')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fcda7a0-c7e3-4ff9-b2de-374d9a857ec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Per-contig metrics files can be very large as reference genomes can contain thousands of alt and decoy contigs.\n",
    "# We exclude those extraneous metrics here.\n",
    "def canonical_chromosome_rows_only(data_rows):\n",
    "    rows_subset = []\n",
    "    for data_index,data_row in data_rows.iterrows():\n",
    "        m = canonical_chr_regex.match(data_row[0])\n",
    "        if m:\n",
    "            #print(\"Kept row with chr\"+m.group(1))\n",
    "            rows_subset.append(data_row)\n",
    "    return pd.DataFrame(rows_subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a40e8bb-0baf-4d85-b85d-7b593f0cc62c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pivot_metric_rows_to_tumor_normal_pair_aggregate_metrics_row(sample_type, data_rows, tumor_normal_pair_metrics):\n",
    "    for data_index,data_row in data_rows.iterrows():\n",
    "        # Make an exception for the histogram CSV files where we only have 2 columns.\n",
    "        if isinstance(data_row[2], float) and math.isnan(data_row[2]):\n",
    "            metric_column_name = sample_type + \" \" + data_row[0]\n",
    "            tumor_normal_pair_metrics[metric_column_name] = data_row[1]\n",
    "        # Make an exception for contig mean coverage metrics files, where there are 3 columns, a raw count ignored in 2, and the fold-coverage in 3\n",
    "        elif isinstance(data_row[2], float) and math.isnan(data_row[3]):\n",
    "            metric_column_name = sample_type + \" \" + data_row[0]\n",
    "            tumor_normal_pair_metrics[metric_column_name] = data_row[2]\n",
    "        # FASTQC metrics get labelled as Read1 or Read2 related.\n",
    "        elif isinstance(data_row[1], str) and data_row[1].startswith(\"Read\"):\n",
    "            metric_column_name = sample_type + \" \" + data_row[0] + \" \" + data_row[1] + \" \" + data_row[2]\n",
    "            tumor_normal_pair_metrics[metric_column_name] = data_row[3]\n",
    "        # We are not going to report per-read group metrics\n",
    "        elif data_row[0].endswith(\"PER RG\"):\n",
    "            continue\n",
    "        else:\n",
    "            metric_column_name = sample_type + \" \" + data_row[0] + \" \" + data_row[2]\n",
    "            value = data_row[3]\n",
    "            # Transform all time intervals into seconds for plotting.\n",
    "            if isinstance(value, str):\n",
    "                m = hhmmss_regex.match(value)\n",
    "                if m:\n",
    "                    value = int(m.group(1))*3600+int(m.group(2))*60+int(m.group(3))\n",
    "            tumor_normal_pair_metrics[metric_column_name] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "738739bd-41b6-4951-96a1-424ee56c2b73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Every subject's tumor-normal pair is allowed to have one value per column.\n",
    "# We only generate stats for the samples that are in the DNA or RNA sample files.\n",
    "for metadata_index,metadata_row in dna_meta.iterrows():\n",
    "    subject = metadata_row[0] \n",
    "    print(subject)\n",
    "    # Pull out the metadata row that corresponds to the subject\n",
    "    tumor = metadata_row[1]\n",
    "    project = metadata_row[9]\n",
    "\n",
    "    tumor_normal_pair_metrics = {}\n",
    "    tumor_normal_pair_metrics[\"subject\"] = subject\n",
    "    tumor_normal_pair_metrics[\"project\"] = project\n",
    "    tumor_normal_pair_metrics[\"tumor\"] = tumor\n",
    "    tumor_normal_pair_metrics[\"TumorHasPCRDuplicates\"] = metadata_row[2]\n",
    "    tumor_normal_pair_metrics[\"PCGRTissue_code\"] = metadata_row[6]\n",
    "    tumor_normal_pair_metrics[\"OncoTree_code\"] = metadata_row[7]\n",
    "    tumor_normal_pair_metrics[\"TGCA_code\"] = metadata_row[8]\n",
    "    # See if there is a germline sample, and load the metrics if they exist.\n",
    "    if metadata_row[3] != \"NO_NORMAL\": \n",
    "        germline = metadata_row[3]\n",
    "        tumor_normal_pair_metrics[\"normal\"] = germline\n",
    "        tumor_normal_pair_metrics[\"NormalHasPCRDuplicates\"] = metadata_row[4]\n",
    "        tumor_normal_pair_metrics[\"normal_contains_tumor\"] = metadata_row[5]\n",
    "        for suffix in dna_germline_csv_metric_file_suffices:\n",
    "            filepath = config[\"output_dir\"]+\"/\"+project+\"/\"+subject+\"/\"+subject+\"_\"+germline+suffix\n",
    "            if os.path.isfile(filepath):\n",
    "                print(\"germline dna: \"+filepath)\n",
    "                # Some metrics files have some rows with 5 columns, where the 5th is a % measure vs an absolute number in the 4th column.\n",
    "                # Passing in 5 names resolves a default problem of parsing a ragged CSV file. \n",
    "                data_rows = pd.read_csv(filepath, sep=\",\", names=range(5), header=None)\n",
    "                pivot_metric_rows_to_tumor_normal_pair_aggregate_metrics_row(\"Normal\", data_rows, tumor_normal_pair_metrics)\n",
    "                    \n",
    "        for suffix in dna_germline_normal_csv_metric_file_suffices:\n",
    "            filepath = config[\"output_dir\"]+\"/\"+project+\"/\"+subject+\"/\"+subject+\"_\"+tumor+\"_\"+germline+suffix\n",
    "            if os.path.isfile(filepath):\n",
    "                print(\"germline paired dna: \"+filepath)\n",
    "                data_rows = pd.read_csv(filepath, sep=\",\", names=range(5), header=None)\n",
    "                pivot_metric_rows_to_tumor_normal_pair_aggregate_metrics_row(\"Normal\", data_rows, tumor_normal_pair_metrics)\n",
    "\n",
    "        for suffix in dna_germline_contig_csv_metric_file_suffices:\n",
    "            filepath = config[\"output_dir\"]+\"/\"+project+\"/\"+subject+\"/\"+subject+\"_\"+germline+suffix\n",
    "            if os.path.isfile(filepath):\n",
    "                print(\"germline dna contigs: \"+filepath)\n",
    "                data_rows = pd.read_csv(filepath, sep=\",\", names=range(5), header=None)\n",
    "                pivot_metric_rows_to_tumor_normal_pair_aggregate_metrics_row(\"Normal mean coverage\", canonical_chromosome_rows_only(data_rows), tumor_normal_pair_metrics)\n",
    "\n",
    "#        for suffix in dna_germline_contig_normal_csv_metric_file_suffices:\n",
    "#            filepath = config[\"output_dir\"]+\"/\"+project+\"/\"+subject+\"/\"+subject+\"_\"+tumor+\"_\"+germline+suffix\n",
    "#            if os.path.isfile(filepath):\n",
    "#                print(\"germline dna paired contigs: \"+filepath)\n",
    "#                data_rows = pd.read_csv(filepath, sep=\",\", names=range(5), header=None)\n",
    "#                pivot_metric_rows_to_tumor_normal_pair_aggregate_metrics_row(\"Normal\", canonical_chromosome_rows_only(data_rows), tumor_normal_pair_metrics)\n",
    "\n",
    "    \n",
    "    # Assume there is a somatic sample, load metrics if they exist. Tumor-only is not yet supported.\n",
    "    for suffix in dna_somatic_csv_metric_file_suffices:\n",
    "        filepath = config[\"output_dir\"]+\"/\"+project+\"/\"+subject+\"/\"+subject+\"_\"+tumor+\"_\"+germline+suffix\n",
    "        if os.path.isfile(filepath):\n",
    "            print(\"somatic paired dna: \"+filepath)\n",
    "            data_rows = pd.read_csv(filepath, sep=\",\", names=range(5), header=None)\n",
    "            pivot_metric_rows_to_tumor_normal_pair_aggregate_metrics_row(\"Tumor\", data_rows, tumor_normal_pair_metrics)\n",
    "    \n",
    "    for suffix in dna_somatic_contig_csv_metric_file_suffices:\n",
    "        filepath = config[\"output_dir\"]+\"/\"+project+\"/\"+subject+\"/\"+subject+\"_\"+tumor+\"_\"+germline+suffix\n",
    "        if os.path.isfile(filepath):\n",
    "            print(\"somatic dna contigs: \"+filepath)\n",
    "            data_rows = pd.read_csv(filepath, sep=\",\", names=range(5), header=None)\n",
    "            pivot_metric_rows_to_tumor_normal_pair_aggregate_metrics_row(\"Tumor mean coverage\", canonical_chromosome_rows_only(data_rows), tumor_normal_pair_metrics)\n",
    "\n",
    "    metrics.append(tumor_normal_pair_metrics)\n",
    "\n",
    "# Only the first RNA sample found in the config file for any given subject normal will be included in the stats.\n",
    "rna_matched = {}\n",
    "for index,metadata_row in rna_meta.iterrows():\n",
    "    subject = metadata_row[0] \n",
    "    print(subject)\n",
    "    project = metadata_row[3]\n",
    "    rna = metadata_row[1]\n",
    "    matching_tumor_dna = row[2]\n",
    "    if matching_tumor_dna in rna_matched.keys(): # Ignore subsequent RNA samples associated with a tumor.\n",
    "        continue\n",
    "    else:\n",
    "        rna_matched[matching_tumor_dna] = True\n",
    "\n",
    "    # Python uses pass-by-reference, so these tumor_normal_pair_metrics updates end up in the final \n",
    "    # table even though append() was called earlier.\n",
    "    tumor_normal_pair_metrics[\"ImmuneDeconvCancerType\"] = metadata_row[4]\n",
    "    tumor_normal_pair_metrics[\"RNAExpressionCohort\"] = metadata_row[5]\n",
    "    for suffix in rna_csv_metric_file_suffices:\n",
    "        filepath = config[\"output_dir\"]+\"/\"+project+\"/\"+subject+\"/rna/\"+subject+\"_\"+rna+suffix\n",
    "        if os.path.isfile(filepath):\n",
    "            print(\"rna: \"+filepath)\n",
    "            data_rows = pd.read_csv(filepath, sep=\",\", names=range(5), header=None)\n",
    "            pivot_metric_rows_to_tumor_normal_pair_aggregate_metrics_row(\"RNA\", data_rows, tumor_normal_pair_metrics)\n",
    "\n",
    "# Specially-formatted files that need their own processing.\n",
    "# .dna.germline.insert-stats.tab, \n",
    "# .dna.somatic.insert-stats.tab, \n",
    "            \n",
    "# .dna.somatic.hrdscore.csv,\n",
    "# .dna.somatic.microsat_output.json, \n",
    "# and .dna.somatic.cnv.vcf.gz (for ModelSource=DEGENERATE_DIPLOID, etc.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51f61db2-95c0-41df-9443-21f4e7be289b",
   "metadata": {},
   "outputs": [],
   "source": [
    "metrics_dataframe = pd.DataFrame.from_dict(metrics)\n",
    "metrics_dataframe.to_csv(\"resources/metrics/dragen.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4e37ddc-344c-4058-a640-226d95e987ad",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60526ff7-753e-4e06-9f78-5718c2aa34b5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
